### PR TITLE
feat: add CI support for typing

### PR DESCRIPTION
**Pull request by rockstar**
This patch introduces mypy as part of the CI process. It adds (or
fixes) type information where it was trivial to do, updates code
where actual bugs were discovered by `mypy` (bugs that may have
also been discovered via tests), and ignores type checking in
`client.py` and `attributes.py`, as those two files warrant a PR
each to make `mypy` happy.

### CODE REVIEW

**Review by davidstanley01 — APPROVED** [review]

File: src/watchful/client.py [hunk 1]
```diff
@@ -83,7 +91,14 @@ def await_port_opening(port: int, timeout_sec: int = 10) -> None:
     raise TimeoutError("Timed out waiting for Watchful to start")
 
 
-def spawn_cmd(cmd: str, env: str = None) -> int:
+def spawn_cmd(
+    cmd: str,
+    env: Union[
+        Mapping[bytes, Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]],
```
Comment (ryttry) [inline src/watchful/client.py:97]: Why not `Optional[...]` like the current style/other changes made?

File: src/watchful/client.py [hunk 2]
```diff
@@ -83,7 +91,14 @@ def await_port_opening(port: int, timeout_sec: int = 10) -> None:
     raise TimeoutError("Timed out waiting for Watchful to start")
 
 
-def spawn_cmd(cmd: str, env: str = None) -> int:
+def spawn_cmd(
+    cmd: str,
+    env: Union[
```
Comment (ryttry) [inline src/watchful/client.py:96]: Why not Optional[...] like the current style/other changes made?

File: src/watchful/client.py (hunk 1 above)
Comment (ryttry) [inline src/watchful/client.py:97]: Why not `Dict`?

File: src/watchful/enricher.py [hunk 3]
```diff
@@ -83,7 +82,13 @@ def __init__(self) -> None:
         class.
         """
 
-        pass
+    @abstractproperty
```
Comment (ryttry) [inline src/watchful/enricher.py:85]: https://docs.python.org/3/library/abc.html#abc.abstractproperty:~:text=%40abc.abstractproperty-,Deprecated%20since%20version%203.3,-%3A%20It%20is

File: src/watchful/client.py (hunk 2 above)
Comment (rockstar) [reply src/watchful/client.py:96]: The `Optional[...]` ones were changed by a tool, and will need to be revisited (most of them are wrong, and being ignored in this patch). In this case, `spawn_cmd` is calling into `subprocess.Popen`, and specifying an `env`. This type now matches the type of `env` in `subprocess.Popen`.

File: src/watchful/client.py (hunk 1 above)
Comment (rockstar) [reply src/watchful/client.py:97]: See above comment.

File: src/watchful/enricher.py (hunk 3 above)
Comment (rockstar) [reply src/watchful/enricher.py:85]: See the `XXX` comment below it: this code exists _solely_ to make `mypy` happy. Currently, there's a code path that references `enrichment_args` but, prior to this patch, `Enricher` didn't have an `enrichment_args`. It still won't work if executed, but this takes us far enough along to address that in a patch that isn't many mechanical changes.

**Pull request merged**